### ISSUE TITLE
Power Cycle Ports

### ISSUE THREAD

**Issue by xaaron** [unifi-poller/unifi#43]
Great tool.  I would like to have the ability to power cycle ports on a POE switch.
Based on https://ubntwiki.com/products/software/unifi-controller/api Something like the following:
func (api *Unifi) PowerCycle(site string, switch_mac string, portIndex int) error {
        manager := "devmgr"
        command := "power-cycle"

        jsonString := fmt.Sprintf(`"data":[{"cmd":"%s","mac":"%s","port_idx":%d}]`, command, switch_mac, portIndex)
        fmt.Println("JSON:  " + jsonString)
        url := fmt.Sprintf("/proxy/network/api/s/%s/cmd/%s", site, manager)
        req := struct {
                Command   string `json:"cmd"`
                Mac       string `json:"mac"`
                PortIndex int    `json:"port_idx"`
        }{
                Command:   command,
                Mac:       switch_mac,
                PortIndex: portIndex,
        }
        return api.postCommand(url, &req, &json.RawMessage{}, reqOpts{})
}

**Comment by davidnewhall** [comment]
The library currently contains no code that "changes things." Not because it shouldn't, but just because I've never needed it. This seems like a very reasonable request and I don't see a problem with adding this.

**Comment by xaaron** [comment]
I'm a newbie at this.  I need your guidance to do this correctly.  This reboots port successfully.  've created a file commands.go (because I assume more commands to unifi will be added) with the following contents:

package unifi
import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
)

func (api *Unifi) PowerCycle(manager string, command string, switch_mac string, portIndex int) {
	manager := "devmgr"
	//command:  "power-cycle"
	//site:  "default" or whatever the site name for switch with port to power cycle
	//switch_mac:  "aa:bb:cc:dd:ee:ff"  mac address of the switch in the site above
	//portIndex:  10 // port number to power cycle.  Value 1-48 for 48 port switch
	req := struct {
		Command   string `json:"cmd"`
		Mac       string `json:"mac"`
		PortIndex int    `json:"port_idx"`
	}{
		Command:   command,
		Mac:       switch_mac,
		PortIndex: portIndex,
	}
	data, err := json.Marshal(req)
	if err != nil {
		panic("internal error marshaling JSON POST body: " + err.Error())
	}
	body := bytes.NewReader(data)
	// Not sure how to derive url with respect to unifi-poller/unifi, udm etc.  Hardcoded and need help here do do this correctly
	url := fmt.Sprintf("https://127.0.0.1:8443/api/s/%s/cmd/%s", "default", manager) //site is "default" and manager is "devmgr"
	resp, err := api.Client.Post(url, "application/json; charset=utf-8", body)
	if err != nil {
		log.Fatal("Unable to post data!")
	}
	log.Printf("ResponseStatus:  %v\n", resp.Status)
	response, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Printf("Error reading response body")
	}
	resp.Body.Close()
	log.Printf("RESPONSE:  %s\n", string(response))
}

**Comment by davidnewhall** [comment]
What's the error?

**Comment by xaaron** [comment]
There is no error from the existing code for older (5.x.X) and non UDM
controllers.  I simply provided the base code in case there is interest in
implementing this in the library, then for newer (6.x.x) and UDM, the
prefix for api calls need to be /proxy/network.  For the newer controllers,
I believe the prefix needs to change, but I am not sure if there is a
mechanism to do that in the existing code.  I've gone ahead and implemented
the below commands in the commands.go file if you are interested in using
it.  I've also extracted out some common code into a 'runCommand(manager
string, site string, option string) function that is common after setting
the json structs.
From the documentation:
UDM Pro API
NOTE: There are two critical differences between Unifi controllers and the
UDM Pro's API:

The login endpoint is /api/auth/login
All API endpoints need to be prefixed with /proxy/network
Callable commands

Posting to the endpoint `api/s/{site}/cmd/<manager>` with the `json {"cmd":
"command"}` you can invoke commands on the controller.
ManagerCallNotes
evtmgt archive-all-alarms
sitemgr add-site desc = Descriptive name ( required ), name = shortname (
in the URL )
sitemgr delete-site name = short name ( required )
sitemgr update-site desc = Descriptive name ( required )
sitemgr get-admins List all administrators and permission for this site
sitemgr move-device mac = device mac ( required ), site_id = 24 digit id (
required )
sitemgr delete-device mac = device mac ( required )
stamgr block-sta mac = client mac ( required )
stamgr unblock-sta mac = client mac ( required )
stamgr kick-sta Disconnect: mac = client mac (required )
stamgr forget-sta Forget a client ( controller 5.9.x only )
stamgr unauthorize-guest Unauthorize a client device, mac = client mac
(required)
devmgr adopt mac = device mac ( required )
devmgr restart mac = device mac ( required )
devmgr force-provision mac = device mac ( required )
devmgr power-cycle mac = switch mac ( required ), port_idx = PoE port to
cycle ( required )
devmgr speedtest Start a speed test
devmgr speedtest-status get the current state of the speed test
devmgr set-locate mac = device mac ( required ) blink unit to locate
devmgr unset-locate mac = device mac ( required ) led to normal state
devmgr upgrade mac = device mac ( required ) upgrade firmware
devmgr upgrade-external mac = device mac ( required ), url = firmware URL (
required )
devmgr migrate mac = device mac ( required ), inform_url = New Inform URL to
push to device (required)
devmgr cancel-migrate mac = device mac ( required )
devmgr spectrum-scan mac = device mac ( ap only, required ) trigger RF scan
backup list-backups list of autobackup files
backup delete-backup filename ( required )
system backup create a backup. This appears to backup to a fixed location
in the filesystem
stat clear-dpi resets the site wide DPI counters

**Comment by davidnewhall** [comment]
Okay I understand. Thank you for providing the base code!

**Comment by davidnewhall** [comment]
Added this. Would love to know if you could test it.

**Comment by xaaron** [comment]
The commands are at:  https://ubntwiki.com/products/software/unifi-controller/api#callable

**Comment by xaaron** [comment]
I had to make the following adjustment in the devmgr.go around line 32 to get this to work:
	 myurl := fmt.Sprintf(APIDevMgrPath, u.SiteName)
	_, err = u.controller.GetJSON(myurl, string(data))

It seems that APIDevMgrPath is defined:
types.go:45:	APICommandPath   string = "/api/s/%s/cmd" // here the site needs to be set with u.SiteName
types.go:46:	APIDevMgrPath    string = APICommandPath + "/devmgr"

**Comment by PianSom** [comment]
@xaaron - please can I ask whether you have used this with a >1.9 firmware on a UDM/P or >=6.1.61 controller? 

The reason for asking is that I have seen many reports complaining that the "one click" functionality for a switch port PoE restart has been removed from the controller. I don't know whether this is just a ui change or a more fundamental API thing. 

I haven't updated yet, and so this could be a very useful addition to poller

**Comment by davidnewhall** [comment]
@PianSom I won't be giving poller the ability to restart switch ports. This is just a base unifi library that poller uses. The library may expose features poller won't use.

@xaaron It looks like I forgot something there. whoops. Does it work OK after adding the `Sprintf`?

**Comment by xaaron** [comment]
Yes, it works great with the Sprintf.

**Comment by davidnewhall** [comment]
Thanks for all the info and the wiki link! I'll get this tidied up shortly.

**Comment by xaaron** [comment]
@PianSom I don't have the newer controller nor the UDM/P in production.  I do use the container provided by https://github.com/jacobalberty/unifi-docker as a Podman container (for unifi-poller also), and will test with newer controller container and potentially a UDM/P once the enhancement is rolled into main.  I plan to test it out in the next month or so.

**Comment by davidnewhall** [comment]
I pushed some more changes to that branch/PR. It includes the `Sprintf` fix as well as adds Device `Adopt` and `Restart`. I also added UAP `ScanRF`.  Can you test the `Restart` for me (it'll restart a device)? You can fire it from `site` or `device` with the `Restart()` methods. If you need more explanation I can provide it, but I suspect you know this code well enough to figure it out. Thank you!

**Comment by davidnewhall** [comment]
I'm also not sure if I handle errors as well I should. Can you try a `site.Restart()` on a MAC that does not exist and tell me if the error it produces is useful? You can share it here. Thank you again!

**Comment by xaaron** [comment]
@davidnewhall I will test this (likely next weekend) and let you know what I find.  

You do a mighty fine job with this library, and I am grateful for your work.  I use this library for monitoring and curating my home automation.  

Previously used homegrown code for this purpose, and as I am new to coding, it wasn't very good code and broke with UDM/P and the new Controller versions, which lead me to this library which got me to unifi-poller/unifi.  

I am happy to help in any way I can.

**Comment by davidnewhall** [comment]
In the meantime I'm probably going to merge that change, and three others. Please open another issue if you run into any problems, and reference this one if it's related. <3
